lesson05/shapes/circle_test.go

```diff
@@ -0,0 +1,151 @@
+package shapes
+
+import (
+	"fmt"
+	"math"
+	"testing"
+)
+
+func TestCircle_SetRadius(t *testing.T) {
+	type fields struct {
+		radius float64
+	}
+	type args struct {
+		r float64
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		args   args
+		want   error
+	}{
+		{"positive radius", fields{radius: 1}, args{r: 1}, nil},
+		{"zero radius", fields{radius: 0}, args{r: -1}, fmt.Errorf("radius needs to be positive and non-zero, got: %.3f", 0.0)},
+		{"negative radius", fields{radius: -1}, args{r: -1}, fmt.Errorf("radius needs to be positive and non-zero, got: %.3f", -1.0)},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := &Circle{
+				radius: tt.fields.radius,
+			}
+			if got := c.SetRadius(tt.args.r); got == nil && tt.want != nil || got != nil && tt.want == nil {
```

Comment:
You shouldn't have these complicated if statements in your tests. Also, you should check your result against what you want, not nil:

```go
if got != tt.want{
     ...
}
```

Same goes for other tests below